### PR TITLE
Copy arrays between CPU and GPU

### LINKED ISSUE #1157: Mix CPU mode and GPU mode in one network
I want to use CPU mode for some functions and GPU mode for others in one network.
When I use a large number of vocabulary in EmbedID about >=1,000,000 words, it consumes large size of memory in GPU. In this situation, I need to use EmbedID in CPU, and to use other functions like LSTMs in GPU.

### PR DESCRIPTION
**Pull request by unnonouno**
I fixed copy function to support copying arrays between CPU and GPU.
fix #1157

### CODE REVIEW

File: chainer/functions/array/copy.py [hunk 1]
```diff
@@ -16,16 +16,36 @@ def check_type_forward(self, in_types):
...
-        return cuda.copy(x[0], out_device=self.out_device),
+        if self.out_device == -1:
+            return cuda.to_cpu(x[0]),
+        else:
+            return cuda.copy(x[0], out_device=self.out_device),
+
+    def backward(self, inputs, grad_outputs):
+        # In this function, `grad_outputs` contains cuda arrays even when
+        # `inputs` only contains numpy arrays.
+        if any(isinstance(x, cuda.ndarray) for x in inputs):
```
Comment (okuta) [inline chainer/functions/array/copy.py]: How about `if isinstance(inputs[0], cuda.ndarray):`?

Comment (unnonouno) [inline chainer/functions/array/copy.py]: Do we need to consider case when parts of inputs are `cuda.ndarray`?

Comment (okuta) [inline chainer/functions/array/copy.py]: Copy function is received one Variable (`len(inputs) == 1`) .

**Comment by okuta** [comment]
Please write document.
LGTM except comments.

File: chainer/functions/array/copy.py (hunk 1 above)
Comment (unnonouno) [inline chainer/functions/array/copy.py]: i see.

**Comment by unnonouno** [comment]
i fixed it

**Pull request merged**